llpc/context/llpcPipelineContext.cpp

```diff
@@ -615,89 +655,123 @@ void PipelineContext::setUserDataNodesTable(Pipeline *pipeline, ArrayRef<Resourc
+      inputAssemblyState.primitiveType = PrimitiveType::Patch;
+      break;
+    default:
+      llvm_unreachable("");
+    }
+
+    inputAssemblyState.patchControlPoints = inputIaState.patchControlPoints;
+    inputAssemblyState.disableVertexReuse = inputIaState.disableVertexReuse;
+    inputAssemblyState.switchWinding = inputIaState.switchWinding;
+
+    rasterizerState.rasterizerDiscardEnable = inputRsState.rasterizerDiscardEnable;
+    rasterizerState.usrClipPlaneMask = inputRsState.usrClipPlaneMask;
+    rasterizerState.provokingVertexMode = static_cast<ProvokingVertexMode>(inputRsState.provokingVertexMode);
+  }
+
+  if (stageMask & (1 << ShaderStageFragment)) {
+    rasterizerState.innerCoverage = inputRsState.innerCoverage;
+    rasterizerState.perSampleShading = inputRsState.perSampleShading;
+    rasterizerState.numSamples = inputRsState.numSamples;
+    rasterizerState.samplePatternIdx = inputRsState.samplePatternIdx;
   }
-  if (inputDsState.stencilTestEnable) {
-    depthStencilState.stencilTestEnable = inputDsState.stencilTestEnable;
-    depthStencilState.stencilCompareOpFront = inputDsState.front.compareOp;
-    depthStencilState.stencilCompareOpBack = inputDsState.back.compareOp;
+
+  if (pipeline)
+    pipeline->setGraphicsState(inputAssemblyState, rasterizerState);
+  if (hasher) {
+    hasher->Update(inputAssemblyState);
+    hasher->Update(rasterizerState);
   }
 
-  pipeline->setDepthStencilState(depthStencilState);
+  if (stageMask & (1 << ShaderStageFragment)) {
+    // Fragment shader is present.
+    const auto &inputDsState = static_cast<const GraphicsPipelineBuildInfo *>(getPipelineBuildInfo())->dsState;
+    DepthStencilState depthStencilState = {};
+    if (inputDsState.depthTestEnable) {
+      depthStencilState.depthTestEnable = inputDsState.depthTestEnable;
+      depthStencilState.depthCompareOp = inputDsState.depthCompareOp;
+    }
+    if (inputDsState.stencilTestEnable) {
+      depthStencilState.stencilTestEnable = inputDsState.stencilTestEnable;
+      depthStencilState.stencilCompareOpFront = inputDsState.front.compareOp;
+      depthStencilState.stencilCompareOpBack = inputDsState.back.compareOp;
+    }
+
+    if (pipeline)
+      pipeline->setDepthStencilState(depthStencilState);
+    if (hasher)
+      hasher->Update(depthStencilState);
+  }
 }
 
 // =====================================================================================================================
-// Set vertex input descriptions in middle-end Pipeline object
+// Set vertex input descriptions in middle-end Pipeline object, or hash them.
 //
-// @param pipeline : Pipeline object
-void PipelineContext::setVertexInputDescriptions(Pipeline *pipeline) const {
+// @param [in/out] pipeline : Middle-end pipeline object; nullptr if only hashing
+// @param [in/out] hasher : Hasher object; nullptr if only setting LGC pipeline state
+void PipelineContext::setVertexInputDescriptions(Pipeline *pipeline, Util::MetroHash64 *hasher) const {
```

Comment:
Same as the color export state.  Two different functions please.

Reply:
I think I did that...
The rationale of the same function is that, when someone adds a new bit of state, it makes it more difficult to forget to add it to the hash.

Reply:
You could have the two entry points that share code behind the scenes.  It is very awkward when you read a line "setVertextInputDescriptions", but it won't actually set anything because pipeline is nullptr.

Also, having the hashing code in PipelineContext is inconsistent with the rest of the hashing code, which is located in its own class.  We should be moving to a single design handling the hashing to make things clearer.  I'm open to which way that is, but I don't want a mixture of designs.